### PR TITLE
feat(messages): add message attachments

### PR DESCRIPTION
**Pull request by pavlzk**
### What this PR does 📖
- add message attachments upload/download
- add message keys

### Which issue(s) this PR fixes 🔨
- Resolve #

### CODE REVIEW

File: src/components/main/compose/mod.rs [hunk 1]
```diff
@@ -95,14 +111,32 @@ pub fn Compose(cx: Scope<Props>) -> Element {
                                     state.write().dispatch(Actions::UpdateConversation(conversation_info));
                                 }
 
+                                if selected_file.is_some() {
+                                    println!("attachments: {:?}", &selected_file);
```
Comment (dariusc93) [inline src/components/main/compose/mod.rs]: may want to use log::info or remove the println altogether

File: src/components/main/compose/mod.rs [hunk 2]
```diff
@@ -95,14 +111,32 @@ pub fn Compose(cx: Scope<Props>) -> Element {
...
+                                    let attachments = selected_file.as_ref().unwrap().to_vec();
+                                    if let Err(_e) = warp::async_block_in_place_uncheck(rg.attach(id, attachments, text_as_vec)) {
```
Comment (dariusc93) [inline src/components/main/compose/mod.rs:120]: Just as a note you the message part of the function is optional (although its not Option<Vec<_>> you can send an attachment without the message here vs send where it requires a message)

**Review by dariusc93 — COMMENTED** [review]
Looks good so far. Left a couple of comments but will be taking a closer look at a couple of things in several hours

**Review by InfamousVague — APPROVED** [review]

**Comment by phillsatellite** [comment]
Found a few things when testing I wanted to mention - 

1. It seems like messages with attachments will only send if you have text typed with it as well. Trying to send a file without adding any text will cause it to not send

https://user-images.githubusercontent.com/93608357/204061131-56f6d5d4-2d6f-4c43-a2b7-d653e7cf5f04.mov

2. Not sure why but it seems like most of the files will be blocked from sending 

Screenshot 2022-11-25 at 5 41 54 PM

3. And one last thing that is more of a question, are we adding file preview in a future pr or should it be added with this pr? 

Screenshot 2022-11-25 at 5 47 18 PM

**Comment by InfamousVague** [comment]
We can have a different PR for the preview / design stuff, for now I think we should just try to get the ability to send all kinds of files on the system then merge this if everything else works as expected.

File: src/components/main/compose/mod.rs [hunk 3]
```diff
@@ -35,6 +38,16 @@ pub fn Compose(cx: Scope<Props>) -> Element {
...
     let show_media = use_state(&cx, || false);
+    let selected_file =  use_state(&cx, || -> Option<Vec<PathBuf>> { None });
+
+    let selected_file_str = &selected_file
+        .clone()
+        .as_ref()
+        .unwrap_or(&vec![])
+        .iter()
+        .map(|f| f.clone().into_os_string().into_string().unwrap())
+        .collect::<Vec<String>>()
```
Comment (InfamousVague) [inline src/components/main/compose/mod.rs:51]: can we unwrap_or_default to be safer

File: src/components/main/compose/mod.rs [hunk 4]
```diff
@@ -35,6 +38,16 @@ pub fn Compose(cx: Scope<Props>) -> Element {
...
     let show_warning = use_state(&cx, || true);
     let show_media = use_state(&cx, || false);
+    let selected_file =  use_state(&cx, || -> Option<Vec<PathBuf>> { None });
+
+    let selected_file_str = &selected_file
+        .clone()
+        .as_ref()
+        .unwrap_or(&vec![])
+        .iter()
+        .map(|f| f.clone().into_os_string().into_string().unwrap())
```
Comment (dariusc93) [inline src/components/main/compose/mod.rs]:
```suggestion
        .map(|f| f.clone().into_os_string().into_string().unwrap_or_default())
```

File: src/components/main/compose/mod.rs (hunk 3 above)
Comment (pavlzk) [reply src/components/main/compose/mod.rs:51]: done by @dariusc93

**Comment by pavlzk** [comment]
> We can have a different PR for the preview / design stuff, for now I think we should just try to get the ability to send all kinds of files on the system then merge this if everything else works as expected.

file type filter removed

**Comment by pavlzk** [comment]
> Found a few things when testing I wanted to mention -
> 
> 1. It seems like messages with attachments will only send if you have text typed with it as well. Trying to send a file without adding any text will cause it to not send
> 
>  sending.image.mov 
> 2. Not sure why but it seems like most of the files will be blocked from sending
> 
> Screenshot 2022-11-25 at 5 41 54 PM
> 
> 3. And one last thing that is more of a question, are we adding file preview in a future pr or should it be added with this pr?
> 
> Screenshot 2022-11-25 at 5 47 18 PM

1 and 2 are fixed

**Pull request merged**